### ISSUE TITLE
Get-Culture and $PSCulture do not (consistently) reflect in-session switches to another culture

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#3831]
Note: Perhaps this is by design (although there's at least a buggy inconsistency), given that there's currently no _PowerShell_ mechanism for _in-session_ switching to another culture. If so, it should at least be documented.

Steps to reproduce
------------------

```powershell
"Initially:`t$((Get-Culture).Name)"
[cultureinfo]::currentculture = 'fr-FR'; "`$PSCulture`:`t$PSCulture"; "Get-Culture:`t$((Get-Culture).name)"; ".NET:`t`t$([cultureinfo]::currentculture.name)"
```

Note: 
*  If you try the above _interactively_, be sure to execute the 2nd line above _on a single line_.
* The output below:
  * assumes that `en-US` is the initial culture at session startup time.
  * differs between Windows and Unix: on Unix, `Get-Culture` reflects the switch, but `$PSCulture` doesn't, whereas on Windows neither does.

Expected behavior
-----------------

```none
en-US
fr-FR
fr-FR
fr-FR
```

Actual behavior
---------------

* macOS and Ubuntu

```none
Initially:	en-US
$PSCulture:	en-US
Get-Culture:	fr-FR
.NET:		fr-FR
```

* Windows

```none
Initially:	en-US
$PSCulture:	en-US
Get-Culture:	en-US
.NET:		fr-FR

```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta (v6.0.0-beta.1) on macOS 10.12.5
PowerShell Core v6.0.0-beta (v6.0.0-beta.1) on Ubuntu 16.04.1 LTS
Windows PowerShell v5.1.14393.1198 on Microsoft Windows 10 Pro (64-bit; v10.0.14393)
```

**Comment by iSazonov** [comment]
PowerShell does not follow the changes in .Net. This is true for changing current directory too.

**Comment by mklement0** [comment]
Well, as you can see, in Core it _does_ reflect .NET-level changes  in `Get-Culture`, but not in `$PSCulture`.

Also, with respect to the current directory: there's justification for PowerShell "doing its own thing", because it has its own system of drives, the majority of which have no .NET counterpart.

The same justification doesn't apply to culture settings, which are shared by both worlds, especially given that cmdlets _do_ already respect in-session switching; e.g.:

```powershell
> [cultureinfo]::currentculture = 'fr-FR'; Get-Date

samedi 20 mai 2017 13:39:08

```

**Comment by iSazonov** [comment]
That means we have to make sure PowerShell "not doing its own thing" for Culture.

**Comment by mklement0** [comment]
@iSazonov - yes, and that condition appears to be met (I haven't dug into the source). Do you know something to the contrary? Or are you just saying that an investigation / official feedback is needed?

**Comment by iSazonov** [comment]
@mklement0 Please see Get-Culture code.

**Comment by mklement0** [comment]
Looking at the source code, it looks like `Get-Culture` simply delegates to `$host.CurrentCulture`, which is consistent with the (updated) findings above.

In the case of a console host, this seems to lead [to `ConsoleHost.cs`](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs#L867), which has the following:

```csharp
        public override System.Globalization.CultureInfo CurrentCulture
        {
            get
            {
                lock (hostGlobalLock)
                {
#if !CORECLR
                    return NativeCultureResolver.Culture;
#else
                    return CultureInfo.CurrentCulture;
#endif
                }
            }
        }
```

* That explains why _Core's_ `Get-Culture` / `$host.CurrentCulture` is always in sync with `[cultureinfo]::CurrentCulture`.

* For _Windows_ PowerShell, the [value is _cached_ in `NativeCultureResolver.cs`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/hostifaces/NativeCultureResolver.cs#L195), which explains why it can get out of sync with the thread's changed-after-session-startup culture.

So we can explain the disparity, but the question is _what the design intent was_ and, therefore,  _how the inconsistency should be resolved_ (my vote is to always reflect the thread's then-current culture).

**Comment by iSazonov** [comment]
I agree that a consistency should be. After quickly view codes, it seems PowerShell use  Cultureinfo.CurrentCulture without caching.

**Comment by iSazonov** [comment]
@mklement0 Could you please update a status of the issue? Should we consider a fix for $PSCulture?

**Comment by iSazonov** [comment]
@SteveL-MSFT We need PowerShell Committee conclusion.

**Comment by mklement0** [comment]
See also: the related discussion in https://github.com/PowerShell/PowerShell/issues/3833#issuecomment-431837446

**Comment by SteveL-MSFT** [comment]
@iSazonov it would help @PowerShell/powershell-committee if you or @mklement0 could summarize the question you want @PowerShell/powershell-committee to answer

**Comment by mklement0** [comment]
@SteveL-MSFT:

In Windows PowerShell:

 * `Get-Culture` reflects _the current user's persistently defined culture_, irrespective of (transient) in-session changes.
 * `$PSCulture` and `$host.CurrentCulture` reflect the same value.

In PowerShell Core:

* `Get-Culture` always reflects the _current thread's_ culture, which may or may not have been changed in-session.

* `$host.CurrentCulture` reflects the same value, whereas `$PSCulture` reflects the culture that was in effect at session-startup time.

The suggestion is to correct this inconsistency and have `$PSCulture` reflect the same value as `Get-Culture` and `$host.CurrentCulture`.

As an aside: The above shows that we've already broken backward compatibility, though it won't typically surface, as changing the culture in-session is not common.
When it comes to introducing `Set-Culture` to PS Core, the difference becomes more problematic - see https://github.com/PowerShell/PowerShell/issues/3833#issuecomment-431837446.

**Comment by BrucePay** [comment]
`$PSCulture` and `$PSUICulture` are implemented as specific subclasses of  `PSVariable` . For example `$PSCulture` is defined in CultureVariable.cs and looks like:
```csharp
    /// <summary>
    /// A variable that represents $PSCulture
    /// </summary>
    internal class PSCultureVariable : PSVariable
    {
        /// <summary>
        /// Constructs an instance of the variable.
        /// </summary>
        internal PSCultureVariable()
            : base(SpecialVariables.PSCulture, true, ScopedItemOptions.ReadOnly | ScopedItemOptions.AllScope,
                   RunspaceInit.DollarPSCultureDescription)
        {
        }

        /// <summary>
        /// Gets or sets the value of the variable.
        /// </summary>
        ///
        public override object Value
        {
            get
            {
                DebuggerCheckVariableRead();
                return System.Globalization.CultureInfo.CurrentCulture.Name;
            }
        }
    }
```

As you can see, it's not a cached value and should reflect the current culture.

**Comment by BrucePay** [comment]
I suspect there's a bug where the special binding for `$PSCulture` is being replaced by a regular `PSVariable`. I'm still investigating to what's going on.

**Comment by BrucePay** [comment]
And yes there is a bug right in the initialization code.  In SessionStateScope.cs:SetVariable(), the code takes the carefully constructed custom-type object, extracts the basic information, creates a new vanilla PSVariable and inserts that into session state instead of using the custom variable. I would guess this has been broken for forever. Fixing it to use the original variable is technically a breaking change but it's probably worth it.

**Comment by xjcl** [comment]
**If anyone is stuck on an older version of PowerShell and wants a workaround for this problem, simply insert a subprocess call, e.g. replacing `(Get-Culture).Name` with `powershell "(Get-Culture).Name")`.**

For example at my workplace, the culture constantly get changed to `en-US` for no apparent reason. But for me it is essential that culture is set to `de-DE` so I can view `;`-separated CSV files in Excel (if the culture is set incorrectly, I would need to close all open Excel workbooks, open up PowerShell, issue a `Set-Culture` command, then re-open all Excel workbooks). Hence I added a task in Task Scheduler to fix this issue, the problem is that `Set-Culture` causes Windows Explorer to do a re-load to sort files alphabetically which is quite jarring. Hence why I need this workaround.

"Buggy" code:

```ps1
while (1) { if ((Get-Culture).Name -ne 'de-DE') { echo "Set-Culture"; Set-Culture de-DE }; sleep 10 }
```

Fixed code:

```ps1
while (1) { if ((powershell "(Get-Culture).Name") -ne 'de-DE') { echo "Set-Culture"; Set-Culture de-DE }; sleep 10 }
```
